src/analysis/retail/monk/mistweaver/normalizers/CastLinkNormalizer.ts

```diff
@@ -0,0 +1,173 @@
+      TALENTS_MONK.RISING_SUN_KICK_TALENT.id,
+      TALENTS_MONK.ENVELOPING_MIST_TALENT.id,
+      SPELLS.ENVELOPING_MIST_TFT.id,
+    ],
+    referencedEventType: [EventType.Cast],
+    forwardBufferMs: 500,
+    backwardBufferMs: 500,
+    anyTarget: true,
+  },
+  // two REMs happen in same timestamp when dancing mists procs
+  {
+    linkRelation: FROM_DANCING_MISTS,
+    linkingEventId: [SPELLS.RENEWING_MIST_HEAL.id],
+    linkingEventType: [EventType.ApplyBuff],
+    referencedEventId: [SPELLS.RENEWING_MIST_HEAL.id],
+    referencedEventType: [EventType.ApplyBuff],
+    anyTarget: true,
+    additionalCondition(linkingEvent, referencedEvent) {
+      return (
+        (linkingEvent as ApplyBuffEvent).targetID !== (referencedEvent as ApplyBuffEvent).targetID
+      );
+    },
+  },
+  // misty peaks proc from a ReM hot event
+  {
+    linkRelation: FROM_MISTY_PEAKS,
+    linkingEventId: [TALENTS_MONK.ENVELOPING_MIST_TALENT.id],
+    linkingEventType: [EventType.ApplyBuff],
+    referencedEventId: SPELLS.RENEWING_MIST_HEAL.id,
+    referencedEventType: [EventType.Heal],
+    anyTarget: true,
+    forwardBufferMs: 50,
+    backwardBufferMs: 50,
+  },
+];
+
+/**
+ * When a spell is cast on a target, the ordering of the Cast and ApplyBuff/RefreshBuff/(direct)Heal
+ * can be semi-arbitrary, making analysis difficult.
+ *
+ * This normalizer adds a _linkedEvent to the ApplyBuff/RefreshBuff/Heal linking back to the Cast event
+ * that caused it (if one can be found).
+ *
+ * This normalizer adds links for Renewing Mist and Enveloping Mist
+ */
+class CastLinkNormalizer extends EventLinkNormalizer {
+  constructor(options: Options) {
+    super(options, EVENT_LINKS);
+  }
+}
+
+/** Returns true iff the given buff application or heal can be matched back to a hardcast */
+export function isFromHardcast(event: AbilityEvent<any>): boolean {
+  if (HasRelatedEvent(event, FROM_RAPID_DIFFUSION) || HasRelatedEvent(event, FROM_MISTY_PEAKS)) {
+    return false;
+  }
+  // 2nd ReM application is the duplicated event
+  if (HasRelatedEvent(event, FROM_DANCING_MISTS)) {
+    if (FOUND_REMS.has(event.timestamp)) {
+      return false;
+    } else {
+      FOUND_REMS.add(event.timestamp);
+    }
+  }
+  if (HasRelatedEvent(event, FROM_HARDCAST)) {
+    return true;
+  }
+  if (HasRelatedEvent(event, BOUNCED)) {
+    const relatedEvents = GetRelatedEvents(event, BOUNCED);
+    let minEvent = relatedEvents[0];
+    // find closest removal/application
+    relatedEvents.forEach(function (rel_ev) {
+      if (
+        Math.abs(rel_ev.timestamp - event.timestamp) <
+        Math.abs(minEvent.timestamp - event.timestamp)
+      ) {
+        minEvent = rel_ev;
+      }
+    });
+    // see if ancestor event can be linked to a hardcast
+    if (minEvent.type === EventType.RemoveBuff) {
+      return isFromHardcast(minEvent as RemoveBuffEvent);
+    } else {
+      return isFromHardcast(minEvent as ApplyBuffEvent);
+    }
+  }
+  return false;
+}
```

Comment:
Big yikes, but I think I understand what you're doing here. Looks like you're resolving the 'double link' problem with logic here, then tracing back bounces if you need to? This seems viable, but I'd like you to add a bit more docs just so future maintainers can figure out the reasoning.

Reply:
Added comments and split out spooky logic to a helper function to explain this one as it is a very necessary step.